### PR TITLE
Add initial cut of build definition goals and general rules to new "Writing VSTS build definitions" doc

### PR DESCRIPTION
**Pull request by mmitche**

### CODE REVIEW

**Comment by mmitche** [comment]
@markwilkie This is what we were chatting about a few days ago.

File: Documentation/WritingBuildDefinitiions.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+## General Goals For Build Definitions
+Build definitions for the .NET Core Ecosystem should be written with the following goals in mind:
+- **A build definition integrates with telemetry systems** - Builds should integrate with telemetry systems to enable a number of scenarios important to building a more effective .NET Core infrastructure ecosystem, including:
+  - Enable build status tracking
+  - Enabe pass/failure analysis
+  - Enable test failure analysis
+  - Enable efficient product code flow via Maestro# and Speculative Package Flow 
+  - Enable efficient asset tracking
+- **A build definition can be used for both OSS and internal development** - To reduce the cost of developing in both the open and closed worlds, sharing the same build definition logic is ideal.  A definition should be able to be run in both OSS and internal contexts and produce results accordingly (e.g. an internal run signs, an OSS run doesn't).
```
Comment (weshaggard) [inline Documentation/WritingBuildDefinitiions.md]: I would add the build definition should be shared between CI and official builds.

File: Documentation/WritingBuildDefinitiions.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+- **A build definition change must be testable in an isolated environment** - A build defniition change should be testable without altering the state of other in-progress changes.
+- **A build definition should branch with the code it serves and maintain its functionality over the servicing lifetime of the code** - Ensure that not only does the build definition not drift from the source it serves, but also does not become non-functional over time due to changes in external systems.
+
+## General Rules
+- **A build definition is defined in code (YAML)** - Defining as much build definition configuration in source ensures a testable, servicable, and versionable infrastructure ecosystem.
+- **A build definition does not rely on input variables in the environment** - Environment variables cause lots of issues in build definitions and should be avoided as a means of passing input to a build where possible.
+  - They may not be obvious to readers of logs.
```
Comment (weshaggard) [inline Documentation/WritingBuildDefinitiions.md]: reproducing is a build is another common problem when relying on the env.

**Comment by mmitche** [comment]
@weshaggard @markwilkie @chcosta updated

File: Documentation/WritingBuildDefinitiions.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Writing VSTS Build Definitions
+The purpose of this document is to detail how VSTS YAML build definitions should be written to integrate into the .NET Core Ecosystem and take advantage of the scaffolding availabe in the Aracde repository.
```
Comment (JohnTortugo) [inline Documentation/WritingBuildDefinitiions.md]: Tipo here: "Aracde".

File: Documentation/WritingBuildDefinitiions.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+- **A build definition results are easy to reproduce locally** - Build definitions, especially when being viewed through the context of a recent build log, should be able to be reproduced locally (if possible).  Ambient state dependencies should be minimized where possible.
+- **A build definition can be serviced and altered globally to maintain consistency across the .NET Core ecosystem** - The number of build definitions is large, and making extensive changes across the ecosystem must be efficient.  For example, changes or rollout of new telemetry should be relatively painless.
+- **A build definition change must be testable in an isolated environment** - A build defniition change should be testable without altering the state of other in-progress changes.
```
Comment (JohnTortugo) [inline Documentation/WritingBuildDefinitiions.md]: Typo here: "defniition"

File: Documentation/WritingBuildDefinitiions.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# Writing VSTS Build Definitions
+The purpose of this document is to detail how VSTS YAML build definitions should be written to integrate into the .NET Core Ecosystem and take advantage of the scaffolding availabe in the Arcade repository.
+
+## General Goals For Build Definitions
+Build definitions for the .NET Core Ecosystem should be written with the following goals in mind:
+- **A build definition integrates with telemetry systems** - Builds should integrate with telemetry systems to enable a number of scenarios important to building a more effective .NET Core infrastructure ecosystem, including:
+  - Enable build status tracking.
+  - Enabe pass/failure analysis.
```
Comment (ericsciple) [inline Documentation/WritingBuildDefinitiions.md]: typo `Enabe` -> `Enable`

**Pull request merged**